fairseq_cli/train.py

```diff
@@ -174,17 +174,17 @@ def train(args, trainer, task, epoch_itr):
             stats = get_training_stats(agg.get_smoothed_values())
             progress.log(stats, tag='train', step=num_updates)
 
-            if (
-                not args.disable_validation
-                and args.save_interval_updates > 0
-                and num_updates % args.save_interval_updates == 0
-                and num_updates > 0
-            ):
-                valid_losses = validate(args, trainer, task, epoch_itr, valid_subsets)
-                checkpoint_utils.save_checkpoint(args, trainer, epoch_itr, valid_losses[0])
-
-            if num_updates >= max_update:
-                break
+        if (
+            not args.disable_validation
+            and args.save_interval_updates > 0
+            and num_updates % args.save_interval_updates == 0
+            and num_updates > 0
+        ):
+            valid_losses = validate(args, trainer, task, epoch_itr, valid_subsets)
+            checkpoint_utils.save_checkpoint(args, trainer, epoch_itr, valid_losses[0])
+
+        if num_updates >= max_update:
+            break
 
     # log end-of-epoch stats
     stats = get_training_stats(agg.get_smoothed_values())
```

Comment:
According to my test, This fix doesn't seems to completely fix the train loss issue.

agg is out of the scope of `with metrics.aggregate() as agg:`. Is it supposed to be like that?

Reply:
The scope just controls which log messages are aggregated. The results can be accessed inside or outside of the scope via `agg`.

The problem before was that `validate` was called within the scope of the training `agg`, so mid-epoch valid metrics would get accumulated under the train one.

> According to my test, This fix doesn't seems to completely fix the train loss issue.

What is the test?

Reply:
Yes, you're right, I need to update this.